// hazelcast-client/lib/core/ReadOnlyLazyList.js
"use strict";
/*
 * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadOnlyLazyList = void 0;
class ReadOnlyLazyListIterator {
    constructor(list) {
        this.index = 0;
        this.list = list;
    }
    next() {
        if (this.index < this.list.size()) {
            return { done: false, value: this.list.get(this.index++) };
        }
        else {
            return { done: true, value: undefined };
        }
    }
}
/**
 * Represents a list of values with lazy deserialization.
 */
class ReadOnlyLazyList {
    /** @internal */
    constructor(array, serializationService) {
        this.internalArray = array;
        this.serializationService = serializationService;
    }
    /**
     * Returns list's element at the specified index.
     *
     * @param index element's index
     * @returns element
     */
    get(index) {
        const dataOrObject = this.internalArray[index];
        if (dataOrObject == null) {
            return undefined;
        }
        if (this.serializationService.isData(dataOrObject)) {
            const obj = this.serializationService.toObject(dataOrObject);
            this.internalArray[index] = obj;
            return obj;
        }
        else {
            return dataOrObject;
        }
    }
    /**
     * Returns the size of the list.
     */
    size() {
        return this.internalArray.length;
    }
    /**
     * Returns an iterator for elements in the list.
     */
    values() {
        return new ReadOnlyLazyListIterator(this);
    }
    /**
     * Returns a slice of the list.
     *
     * @param start The beginning of the specified portion of the list (inclusive).
     * @param end The end of the specified portion of the list (exclusive).
     */
    slice(start, end) {
        return new ReadOnlyLazyList(this.internalArray.slice(start, end), this.serializationService);
    }
    /**
     * Returns an array that contains all elements of this list in proper sequence.
     */
    toArray() {
        const arr = [];
        const iterator = this.values();
        for (let item = iterator.next(); !item.done; item = iterator.next()) {
            arr.push(item.value);
        }
        return arr;
    }
    [Symbol.iterator]() {
        return this.values();
    }
}
exports.ReadOnlyLazyList = ReadOnlyLazyList;
