// hazelcast-client/lib/aggregation/AggregatorFactory.js
"use strict";
/*
 * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/** @ignore */ /** */
Object.defineProperty(exports, "__esModule", { value: true });
exports.aggregatorFactory = exports.NUMBER_AVG = exports.MIN = exports.MAX = exports.LONG_SUM = exports.LONG_AVG = exports.INT_SUM = exports.INT_AVG = exports.FLOATING_POINT_SUM = exports.FIXED_SUM = exports.DOUBLE_SUM = exports.DOUBLE_AVG = exports.COUNT = exports.AGGREGATOR_FACTORY_ID = void 0;
const Aggregator_1 = require("./Aggregator");
const core_1 = require("../core");
/** @internal */
exports.AGGREGATOR_FACTORY_ID = -29;
// export const BIG_DECIMAL_AVG = 0; // not implemented in node.js
// export const BIG_DECIMAL_SUM = 1; // not implemented in node.js
// export const BIG_INT_AVG = 2; // not implemented in node.js
// export const BIG_INT_SUM = 3; // not implemented in node.js
/** @internal */
exports.COUNT = 4;
// export const DISTINCT = 5; // returns java serializable, not usable in node.js
/** @internal */
exports.DOUBLE_AVG = 6;
/** @internal */
exports.DOUBLE_SUM = 7;
/** @internal */
exports.FIXED_SUM = 8;
/** @internal */
exports.FLOATING_POINT_SUM = 9;
/** @internal */
exports.INT_AVG = 10;
/** @internal */
exports.INT_SUM = 11;
/** @internal */
exports.LONG_AVG = 12;
/** @internal */
exports.LONG_SUM = 13;
/** @internal */
exports.MAX = 14;
/** @internal */
exports.MIN = 15;
/** @internal */
exports.NUMBER_AVG = 16;
// export const MAX_BY = 17; // needs object to implement Java's Comparable interface
// export const MIN_BY = 18; // needs object to implement Java's Comparable interface
const idToConstructor = {
    [exports.COUNT]: Aggregator_1.CountAggregator,
    [exports.DOUBLE_AVG]: Aggregator_1.DoubleAverageAggregator,
    [exports.DOUBLE_SUM]: Aggregator_1.DoubleSumAggregator,
    [exports.FIXED_SUM]: Aggregator_1.FixedPointSumAggregator,
    [exports.FLOATING_POINT_SUM]: Aggregator_1.FloatingPointSumAggregator,
    [exports.INT_AVG]: Aggregator_1.IntegerAverageAggregator,
    [exports.INT_SUM]: Aggregator_1.IntegerSumAggregator,
    [exports.LONG_AVG]: Aggregator_1.LongAverageAggregator,
    [exports.LONG_SUM]: Aggregator_1.LongSumAggregator,
    [exports.MAX]: Aggregator_1.MaxAggregator,
    [exports.MIN]: Aggregator_1.MinAggregator,
    [exports.NUMBER_AVG]: Aggregator_1.NumberAverageAggregator,
};
/** @internal */
function aggregatorFactory(classId) {
    try {
        return new idToConstructor[classId]();
    }
    catch (e) {
        throw new core_1.HazelcastError('There is no known aggregator with type id ' + classId, e);
    }
}
exports.aggregatorFactory = aggregatorFactory;
